src/cpp/rtps/messages/MessageReceiver.cpp

```diff
@@ -406,6 +419,47 @@ bool MessageReceiver::readSubmessageHeader(CDRMessage_t* msg, SubmessageHeader_t
     return true;
 }
 
+bool MessageReceiver::willAReaderAcceptMsgDirectedTo(const EntityId_t & readerID)
+{
+    if(AssociatedReaders.empty()) {
+        logWarning(RTPS_MSG_IN,IDSTRING"Data received when NO readers are listening");
+        return false;
+    }
+
+    if (readerID != c_EntityId_Unknown) {
+        const auto readers = AssociatedReaders.find(readerID);
+        if (readers != AssociatedReaders.end() && !readers->second.empty()) {
+            return true;
+        }
+    }
+    else {
+        for(const auto & readers : AssociatedReaders) {
+            if (readers.second.empty()) continue;
+            for (const auto & it : readers.second) {
+                if (it->m_acceptMessagesToUnknownReaders) return true;
+            }
+        }
+    }
+
+    logWarning(RTPS_MSG_IN, IDSTRING"No Reader accepts this message (directed to: " <<readerID << ")");
+    return false;
+}
+
+void MessageReceiver::findAllReaders(const EntityId_t & readerID, std::function<void(RTPSReader*)> callback)
+{
+    if (readerID != c_EntityId_Unknown) {
+        const auto readers = AssociatedReaders.find(readerID);
+        for (const auto & it : readers->second) callback(it);
```

Comment:
`readers` should be checked to be different from AssociatedReaders.end() before accessing `->second`